File: apps/windows/TeslaSync.App/shared-surfaces/SkipToContent.cs
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Automation;
using Microsoft.UI.Xaml.Controls;
using TeslaSync.App.Components.UI;
using TeslaSync.App.Core;
using TeslaSync.App.Core.Notifications;

namespace TeslaSync.App.SharedSurfaces;

/// <summary>
/// The native WinUI 3 <c>SkipToContent</c> shared surface — a parity port of
/// web/src/components/feedback/SkipToContent.tsx (WCAG 2.4.1, Bypass Blocks, Level A). It is the keyboard
/// bypass-blocks affordance: a single focusable link, mounted as the very first interactive element of the
/// shell, that is visually hidden until it receives keyboard focus and then reveals a Fluent chip in the
/// top-left; activating it moves focus to (and brings into view) the page's main-content landmark so keyboard /
/// Narrator users do not have to tab through the whole navigation on every page. It composes the shared
/// <see cref="TsButton"/> primitive (the native counterpart of the web <c>VisuallyHidden as="a" focusable</c>
/// skip link), binds the <see cref="SkipToContentViewModel"/> (over the <see cref="ILocalizer"/> i18n facade and
/// the <see cref="ISkipTarget"/> landmark seam), and routes activation through the holder. Because the web source
/// has no data fetch there is no loading / empty / error / stale / offline chrome; the surface's states are the
/// resting (hidden) link, the focused (revealed) link, and the activation outcome (focus moved, or a safe no-op
/// when no landmark is present — the web <c>if (main)</c> guard). The reveal is an instant opacity change with no
/// animation, so it honours the reduce-motion preference by construction. The link carries the localized
/// accessible name and the <c>skip-to-content</c> automation id, and the surface emits the <c>view.opened</c>
/// diagnostic once on <see cref="FrameworkElement.Loaded"/>.
/// </summary>
public sealed partial class SkipToContent : ContentControl, IDisposable
{
    private const double RevealedOpacity = 1.0;
    private const double HiddenOpacity = 0.0;
    private const double EdgeInset = 16.0;

    private readonly SkipToContentViewModel _viewModel;
    private readonly TsButton _link = new()
    {
        Variant = ButtonVariant.Primary,
        HorizontalAlignment = HorizontalAlignment.Left,
        VerticalAlignment = VerticalAlignment.Top,
        Margin = new Thickness(EdgeInset),
        Opacity = HiddenOpacity,
    };

    private bool _disposed;

    /// <summary>
    /// Creates the surface with no composition root (the designer / parameterless host entry point): it binds a
    /// passthrough localizer and an inert landmark seam so the surface renders its resting state. Supply an
    /// explicit <see cref="ILocalizer"/> and a bound <see cref="ISkipTarget"/> via the other constructors to
    /// drive i18n and the main-content landmark from the composition root.
    /// </summary>
    public SkipToContent()
        : this(new SkipToContentViewModel(PassthroughLocalizer.Instance, NullSkipTarget.Instance))
    {
    }

    /// <summary>Creates the surface over the i18n facade and a bound landmark seam (the production entry point).</summary>
    /// <param name="localizer">The i18n facade the label resolves through (web <c>useTranslation</c>).</param>
    /// <param name="target">The main-content landmark seam (web <c>document.getElementById('main-content')</c>).</param>
    /// <param name="diagnostics">The PII-safe diagnostics sink for the <c>view.opened</c> event.</param>
    public SkipToContent(
        ILocalizer localizer,
        ISkipTarget target,
        SkipToContentDiagnostics? diagnostics = null)
        : this(new SkipToContentViewModel(localizer, target, diagnostics))
    {
    }

    /// <summary>Creates the surface over an explicit state holder (tests / headless hosts).</summary>
    /// <param name="viewModel">The backing state holder.</param>
    public SkipToContent(SkipToContentViewModel viewModel)
    {
        ArgumentNullException.ThrowIfNull(viewModel);

        _viewModel = viewModel;

        IsTabStop = false;
        HorizontalAlignment = HorizontalAlignment.Left;
        VerticalAlignment = VerticalAlignment.Top;
        Padding = new Thickness(0);

        _link.Text = _viewModel.Label;

        // The link is the accessible, focusable bypass-blocks control: name it and give it the web data-testid so
        // Narrator and UI automation resolve it.
        AutomationProperties.SetName(_link, _viewModel.Label);
        AutomationProperties.SetAutomationId(_link, SkipToContentRegistration.LinkAutomationId);

        // Visually hidden until focused, revealed on focus (the web `focus:` utilities); an instant opacity
        // change keeps it reduce-motion-safe.
        _link.GotFocus += OnLinkGotFocus;
        _link.LostFocus += OnLinkLostFocus;
        _link.Click += OnLinkClick;

        Loaded += OnLoaded;
        Unloaded += OnUnloaded;

        Content = _link;
    }

    /// <summary>The canonical surface slug (<c>SkipToContent</c>).</summary>
    public static string Slug => SkipToContentRegistration.Slug;

    /// <summary>The backing state holder (exposed for hosting / diagnostics / tests).</summary>
    public SkipToContentViewModel ViewModel => _viewModel;

    /// <inheritdoc />
    public void Dispose()
    {
        if (_disposed)
        {
            return;
        }

        _disposed = true;
        _link.GotFocus -= OnLinkGotFocus;
        _link.LostFocus -= OnLinkLostFocus;
        _link.Click -= OnLinkClick;
        Loaded -= OnLoaded;
        Unloaded -= OnUnloaded;
        GC.SuppressFinalize(this);
    }

    private void OnLoaded(object sender, RoutedEventArgs e) => _viewModel.MarkOpened();

    private void OnUnloaded(object sender, RoutedEventArgs e) => Dispose();

    private void OnLinkGotFocus(object sender, RoutedEventArgs e) => _link.Opacity = RevealedOpacity;

    private void OnLinkLostFocus(object sender, RoutedEventArgs e) => _link.Opacity = HiddenOpacity;

    private void OnLinkClick(object sender, RoutedEventArgs e) => _ = _viewModel.Activate();
}

/// <summary>
/// The production <see cref="ISkipTarget"/> — wraps the WinUI main-content landmark control. Reports
/// <see cref="IsAvailable"/> true and moves focus to the control (the web <c>main.focus(...)</c>) while bringing
/// it into view (the web <c>main.scrollIntoView(...)</c>). The composition root constructs one over the shell's
/// main-content region; when no region exists it uses <see cref="NullSkipTarget"/> instead.
/// </summary>
public sealed class ControlSkipTarget : ISkipTarget
{
    private readonly Control _landmark;

    /// <summary>Creates the adapter over the main-content landmark control.</summary>
    /// <param name="landmark">The control focus jumps to (the shell's main-content region).</param>
    public ControlSkipTarget(Control landmark)
    {
        ArgumentNullException.ThrowIfNull(landmark);
        _landmark = landmark;
    }

    /// <inheritdoc />
    public bool IsAvailable => true;

    /// <inheritdoc />
    public void Focus()
    {
        _ = _landmark.Focus(FocusState.Programmatic);
        _landmark.StartBringIntoView();
    }
}

File: apps/android/app/src/main/java/com/teslasync/shared-surfaces/SkipToContent/SkipToContent.kt
// The native Jetpack Compose + Material 3 SkipToContent shared surface — a parity port of
// web/src/components/feedback/SkipToContent.tsx, the WCAG 2.4.1 (Bypass Blocks) skip link.
//
// The web source composes `<VisuallyHidden as="a" focusable>` with the localized label
// `t('a11y.skipToContent')`: it is invisible to sighted users, reveals itself when it gains keyboard focus
// (`focus:not-sr-only`), and on activation moves focus + scroll to the page's `<main id="main-content">`
// landmark so a keyboard user skips the 50-plus-item sidebar on every page. It MUST be mounted as the very
// first interactive element so one Tab press surfaces it before any sidebar / header control.
//
// [SkipToContent] reproduces that natively: one focusable, clickable `Role.Button` node that is
// layout-negligible while resting (the native analogue of `sr-only`, still carrying the label in the
// accessibility tree so TalkBack always reads it) and reveals a high-contrast chip while focused (the native
// analogue of `focus:not-sr-only`). On activation it routes through the [SkipTarget] seam to focus the
// host's main-content landmark — and a focused node is brought into view inside a scrollable parent, which
// reproduces the web `scrollIntoView`. The bare a11y primitive (the sibling VisuallyHidden surface) carries
// no activation handler, so this surface adds the focus-move the web `onClick` performs while keeping the same
// hidden-until-focus pattern. Mount it first in the host scaffold, and mark the main content once with
// [Modifier.mainContentAnchor].
//
// The view performs no work of its own: it resolves the i18n label (P1/S10) at the render boundary and routes
// activation + the one-shot `view.opened` diagnostic (P1/S11) through [SkipToContentViewModel] over the
// [SkipTarget] seam (P1/S8). The surface fetches nothing, so it has no loading / empty / error / stale /
// offline lifecycle — its real states are the resting [SkipLinkMode.Hidden] link and the focus-revealed
// [SkipLinkMode.Revealed] chip, both reproduced here.
//
// `InvalidPackageDeclaration` is suppressed: the mandated surface directory
// (com/teslasync/shared-surfaces/SkipToContent) cannot form a valid Kotlin package. `MatchingDeclarationName`
// is suppressed for the co-located stateless content, anchor, and previews.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.sharedsurfaces.skiptocontent

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.focusable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsFocusedAsState
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.focus.FocusRequester
import androidx.compose.ui.focus.focusRequester
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.lifecycle.viewmodel.compose.viewModel
import io.teslasync.android.R
import io.teslasync.android.components.ui.BodyText
import io.teslasync.android.data.LocalDataContainer
import io.teslasync.android.ui.theme.TeslaSyncTheme
import io.teslasync.android.ui.theme.generated.Radius
import io.teslasync.android.ui.theme.generated.Spacing
import io.teslasync.shared.core.diagnostics.Logger

/** Test tag identifying the skip link — the native analogue of the web `data-testid="skip-to-content"`. */
const val SKIP_TO_CONTENT_TEST_TAG: String = "skip-to-content"

/** A layout-negligible footprint for the resting (sr-only) skip link — invisible yet present in the a11y tree. */
private val SKIP_LINK_HIDDEN_SIZE = 1.dp

/**
 * Stateful entry point — the parity port of the web `<SkipToContent />`. Binds the [target] seam through
 * [viewModel], records the one-shot `view.opened` diagnostic (P1/S11) on first composition, resolves the
 * localized label (web `t('a11y.skipToContent')`), and renders the focus-revealed skip link. Mount it as the
 * first interactive element in the host scaffold so one Tab press surfaces it (web "mount first" contract).
 *
 * @param target the main-content landmark seam; defaults to the process-wide [ProcessSkipTarget].
 * @param logger the sanctioned redacting logger; defaults to the app's [LocalDataContainer].
 */
@Composable
fun SkipToContent(
    modifier: Modifier = Modifier,
    target: SkipTarget = ProcessSkipTarget,
    logger: Logger = LocalDataContainer.current.logger,
) {
    val viewModel: SkipToContentViewModel =
        viewModel(
            key = SkipToContentRegistration.ID,
            factory = SkipToContentViewModel.factory(target, logger),
        )
    LaunchedEffect(viewModel) { viewModel.onViewOpened() }
    val label = stringResource(R.string.translation_a11y_skipToContent)
    SkipToContentContent(label = label, onActivate = viewModel::skipToContent, modifier = modifier)
}

/**
 * Stateless renderer — the test / preview entry point. Draws the skip link as one focusable, clickable
 * `Role.Button` node whose accessible label is [label] in every state (so TalkBack reads it even while the
 * link is visually hidden). The node is layout-negligible while resting ([SkipLinkMode.Hidden], web `sr-only`)
 * and reveals a high-contrast chip while focused ([SkipLinkMode.Revealed], web `focus:not-sr-only`); activating
 * it fires [onActivate] (web `onClick`).
 */
@Composable
fun SkipToContentContent(
    label: String,
    onActivate: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val interaction = remember { MutableInteractionSource() }
    val focused by interaction.collectIsFocusedAsState()
    val node =
        modifier
            .testTag(SKIP_TO_CONTENT_TEST_TAG)
            .semantics { contentDescription = label }
            .clickable(
                interactionSource = interaction,
                indication = null,
                onClickLabel = label,
                role = Role.Button,
                onClick = onActivate,
            )
    when (skipLinkMode(focused)) {
        SkipLinkMode.Revealed -> RevealedSkipLink(label = label, modifier = node)
        SkipLinkMode.Hidden -> Box(node.size(SKIP_LINK_HIDDEN_SIZE))
    }
}

/**
 * The visible appearance of the focused skip link — a high-contrast chip carrying [label]. Reused by
 * [SkipToContentContent] while focused and exercised directly by the previews, mirroring the sibling
 * VisuallyHidden surface's focus-revealed chip so the two a11y affordances look identical.
 */
@Composable
private fun RevealedSkipLink(
    label: String,
    modifier: Modifier = Modifier,
) {
    Box(
        modifier
            .background(MaterialTheme.colorScheme.inverseSurface, RoundedCornerShape(Radius.sm))
            .padding(horizontal = Spacing.md, vertical = Spacing.sm),
    ) {
        BodyText(label, color = MaterialTheme.colorScheme.inverseOnSurface)
    }
}

/**
 * Marks the receiver as the host's main-content landmark — the native analogue of
 * `<main id="main-content" tabIndex={-1}>`. Registers a focus action with [target] for as long as the node is
 * composed (clearing it on dispose so a torn-down screen never stays the destination) and makes the node a
 * focus target, so a [SkipToContent] activation moves focus here (and brings it into view inside a scrollable
 * parent, reproducing the web `scrollIntoView`). Apply once to the scaffold's main content slot.
 *
 * @param target the registry the landmark registers with; defaults to the process-wide [ProcessSkipTarget].
 */
@Composable
fun Modifier.mainContentAnchor(target: RegistrySkipTarget = ProcessSkipTarget): Modifier {
    val focusRequester = remember { FocusRequester() }
    DisposableEffect(target, focusRequester) {
        val handle = target.register { runCatching { focusRequester.requestFocus() } }
        onDispose { handle.release() }
    }
    return this
        .focusRequester(focusRequester)
        .focusable()
}

// ── Previews — the resting link is invisible by design (asserted in tests), so these render the focus-revealed
// chip in light and dark, mirroring the sibling VisuallyHidden surface. ───────────────────────────────────────

@Preview(name = "SkipToContent — revealed on focus", showBackground = true)
@Composable
private fun SkipToContentRevealedPreview() {
    TeslaSyncTheme(dynamicColor = false) {
        RevealedSkipLink(label = "Skip to main content")
    }
}

@Preview(name = "SkipToContent — revealed (dark)", showBackground = true)
@Composable
private fun SkipToContentRevealedDarkPreview() {
    TeslaSyncTheme(darkTheme = true, dynamicColor = false) {
        RevealedSkipLink(label = "Skip to main content")
    }
}
